Port WebExtension Page and Content Security Properties to C++

ItsJamie9494 opened the pull request:
#### ea61e1921e0f26dfcba16ccc16e239229dda259d
<pre>
Port WebExtension Page and Content Security Properties to C++
<a href="https://webkit.org/b/279719">https://webkit.org/b/279719</a>

Reviewed by Timothy Hatcher.

Ports the Page and Content Security strings including Options and Overrides to C++.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::parseManifest):
(WebKit::WebExtension::createError):
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded):
(WebKit::WebExtension::contentSecurityPolicy): Deleted.
(WebKit::WebExtension::populateContentSecurityPolicyStringsIfNeeded): Deleted.
(WebKit::WebExtension::hasOptionsPage): Deleted.
(WebKit::WebExtension::hasOverrideNewTabPage): Deleted.
(WebKit::WebExtension::optionsPagePath): Deleted.
(WebKit::WebExtension::overrideNewTabPagePath): Deleted.
(WebKit::WebExtension::populatePagePropertiesIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::contentSecurityPolicy):
(WebKit::WebExtension::populateContentSecurityPolicyStringsIfNeeded):
(WebKit::WebExtension::hasOptionsPage):
(WebKit::WebExtension::hasOverrideNewTabPage):
(WebKit::WebExtension::optionsPagePath):
(WebKit::WebExtension::overrideNewTabPagePath):
(WebKit::WebExtension::populatePagePropertiesIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::createError):

Canonical link: <a href="https://commits.webkit.org/283676@main">https://commits.webkit.org/283676@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e80fa7e1c096c50e085242e6d00aeaf4edc64f66

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67024 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46399 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19646 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71059 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18157 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17941 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53756 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12219 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70091 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42676 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57991 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34275 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39347 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16511 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61276 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15726 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72760 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10981 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15064 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61225 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11013 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58050 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61301 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9020 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2626 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10170 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/42206 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43283 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44466 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43026 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -2146,7 +2045,7 @@ static bool parseCommandShortcut(const String& shortcut, OptionSet<ModifierFlags
         auto optionalRuleset = parseDeclarativeNetRequestRulesetDictionary(rulesetDictionary, &error);
         if (!optionalRuleset) {
-            recordError(createError(Error::InvalidDeclarativeNetRequest, nil, error));
+            recordError(createError(Error::InvalidDeclarativeNetRequest, String(), error));
```
```suggestion
            recordError(createError(Error::InvalidDeclarativeNetRequest, nullString(), error));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -221,14 +214,14 @@
     m_manifest = parseJSON(manifestData, { }, &parseError);
     if (!m_manifest) {
-        recordError(createError(Error::InvalidManifest, nil, parseError));
+        recordError(createError(Error::InvalidManifest, String(), parseError));
```
```suggestion
        recordError(createError(Error::InvalidManifest, nullString(), parseError));
```
Or:
```suggestion
        recordError(createError(Error::InvalidManifest, { }, parseError));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -221,14 +214,14 @@
     auto manifestJSON = JSON::Value::parseJSON(manifestJSONString);
     if (!manifestJSON || !manifestJSON->asObject()) {
-        recordError(createError(Error::InvalidManifest, nil, nil));
+        recordError(createError(Error::InvalidManifest, String(), nil));
```
```suggestion
        recordError(createError(Error::InvalidManifest, nullString(), nil));
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -133,6 +143,104 @@ void WebExtension::populateDisplayStringsIfNeeded()
+    return m_contentSecurityPolicy;
+}
+
+
```
```suggestion

```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -133,6 +143,104 @@ void WebExtension::populateDisplayStringsIfNeeded()
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/content_security_policy
+
+    if (supportsManifestVersion(3)) {
+        if (auto policyObject = manifestObject->getObject(contentSecurityPolicyManifestKey)) {
```
```suggestion
        if (RefPtr policyObject = manifestObject->getObject(contentSecurityPolicyManifestKey)) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -133,6 +143,104 @@ void WebExtension::populateDisplayStringsIfNeeded()
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/options_page
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/chrome_url_overrides
+
+    auto optionsDictionary = manifestObject->getObject(optionsUIManifestKey);
```
```suggestion
    RefPtr optionsDictionary = manifestObject->getObject(optionsUIManifestKey);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -133,6 +143,104 @@ void WebExtension::populateDisplayStringsIfNeeded()
+            recordError(createError(Error::InvalidOptionsPage));
+    }
+
+    auto overridesDictionary = manifestObject->getObject(browserURLOverridesManifestKey);
```
```suggestion
    RefPtr overridesDictionary = manifestObject->getObject(browserURLOverridesManifestKey);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -133,6 +143,103 @@ void WebExtension::populateDisplayStringsIfNeeded()
+
+    RefPtr optionsDictionary = manifestObject->getObject(optionsUIManifestKey);
+    if (optionsDictionary) {
+        m_optionsPagePath = optionsDictionary->getString(optionsUIPageManifestKey);
```
Sorry didn't catch this earlier but …Object would be a better suffix for these now.
```suggestion
    if (RefPtr optionsObject = manifestObject->getObject(optionsUIManifestKey)) {
        m_optionsPagePath = optionsObject->getString(optionsUIPageManifestKey);
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -133,6 +143,103 @@ void WebExtension::populateDisplayStringsIfNeeded()
+
+    RefPtr overridesDictionary = manifestObject->getObject(browserURLOverridesManifestKey);
+    if (!overridesDictionary)
+        overridesDictionary = manifestObject->getObject(chromeURLOverridesManifestKey);
```
```suggestion
    RefPtr overridesObject = manifestObject->getObject(browserURLOverridesManifestKey);
    if (!overridesObject)
        overridesObject = manifestObject->getObject(chromeURLOverridesManifestKey);
```

ItsJamie9494 committed:
Port WebExtension Page and Content Security Properties to C++
https://webkit.org/b/279719

Reviewed by Timothy Hatcher.

Ports the Page and Content Security strings including Options and Overrides to C++.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::parseManifest):
(WebKit::WebExtension::createError):
(WebKit::WebExtension::populateDeclarativeNetRequestPropertiesIfNeeded):
(WebKit::WebExtension::contentSecurityPolicy): Deleted.
(WebKit::WebExtension::populateContentSecurityPolicyStringsIfNeeded): Deleted.
(WebKit::WebExtension::hasOptionsPage): Deleted.
(WebKit::WebExtension::hasOverrideNewTabPage): Deleted.
(WebKit::WebExtension::optionsPagePath): Deleted.
(WebKit::WebExtension::overrideNewTabPagePath): Deleted.
(WebKit::WebExtension::populatePagePropertiesIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::contentSecurityPolicy):
(WebKit::WebExtension::populateContentSecurityPolicyStringsIfNeeded):
(WebKit::WebExtension::hasOptionsPage):
(WebKit::WebExtension::hasOverrideNewTabPage):
(WebKit::WebExtension::optionsPagePath):
(WebKit::WebExtension::overrideNewTabPagePath):
(WebKit::WebExtension::populatePagePropertiesIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::createError):

Canonical link: https://commits.webkit.org/283676@main
